# <img style="float: left; padding-right: 20px; width: 200px" src="https://raw.githubusercontent.com/raxlab/imt2200-data/main/media/logo.jpg">  IMT 2200 - Introducción a Ciencia de Datos
**Pontificia Universidad Católica de Chile**<br>
**Instituto de Ingeniería Matemática y Computacional**<br>
**Semestre 2025-S2**<br>
**Profesor:** Rodrigo A. Carrasco <br>

# <h1><center>Actividad 05: Limpieza y Transformación de Datos</center></h1>

Esta actividad busca aplicar los conocimientos en transformación de datos para contestar algunas preguntas.

## Instrucciones

Esto Notebook contiene las instrucciones a realizar para la actividad. 

<b>Al finalizarla, deben subir el Notebook y los archivos generados en un único archivo .zip, al módulo de la Actividad 05 en Canvas.</b>

Para esta actividad trabajarán con el archivo `spotify_history.csv`, que contiene el historial de actividad de un usuario de Spotify durante varios años.

## Rúbrica

- Si han logrado todo (con errores menores): 7.0
- Si han logrado los puntos 1 y 2 (con errores menores): 5.0
- Si no han logrado hasta el punto 2.0: 1.0


## 1. Lectura y selección de columnas

a. Primero, lea el `spotify_history.csv` con Pandas como un único DataFrame y muestre todas las columnas del archivo original.

b. Para esta actividad, las columnas de nuestro interés son las siguientes:
```
['ts','ms_played',
'master_metadata_track_name', 'master_metadata_album_artist_name',
'master_metadata_album_album_name', 'episode_name',
'episode_show_name',
'reason_start', 'reason_end']
```
Seleccione solamente estas columnas para su DataFrame de trabajo.

c. Responda:
* ¿Cuántas filas tiene su dataset?

In [21]:
# a.
import pandas as pd
df = pd.read_csv("spotify_history.csv")
df.head()


ts platform  ms_played conn_country  \
0  2022-12-29T16:39:18Z  android        753           CL   
1  2022-12-29T16:39:19Z  android       1661           CL   
2  2022-12-29T16:39:21Z  android       1000           CL   
3  2022-12-29T16:39:22Z  android       1097           CL   
4  2022-12-29T16:43:38Z  android     256920           CL   

  master_metadata_track_name master_metadata_album_artist_name  \
0               Warning Sign                          Coldplay   
1                    Houdini                 Foster The People   
2               Fingerprints                    Hiatus Kaiyote   
3            Waves of Nature                            WILLOW   
4               Fingerprints                    Hiatus Kaiyote   

  master_metadata_album_album_name                     spotify_track_uri  \
0      A Rush of Blood to the Head  spotify:track:4bPkBHKLKd9WHizsvM2zV3   
1                          Torches  spotify:track:5Fli1xRi01bvCjsZvKWro0   
2               Choose Your Weapon  spotify:track:3c7N2fCFB4JVKGF8cWfteh   
3                     ARDIPITHECUS  spotify:track:1zJMFf0EzbGCpL0zROfmVF   
4               Choose Your Weapon  spotify:track:3c7N2fCFB4JVKGF8cWfteh   

  episode_name episode_show_name  ... audiobook_title  audiobook_uri  \
0          NaN               NaN  ...             NaN            NaN   
1          NaN               NaN  ...             NaN            NaN   
2          NaN               NaN  ...             NaN            NaN   
3          NaN               NaN  ...             NaN            NaN   
4          NaN               NaN  ...             NaN            NaN   

   audiobook_chapter_uri  audiobook_chapter_title  reason_start reason_end  \
0                    NaN                      NaN        fwdbtn     fwdbtn   
1                    NaN                      NaN        fwdbtn     fwdbtn   
2                    NaN                      NaN        fwdbtn     fwdbtn   
3                    NaN                      NaN        fwdbtn    backbtn   
4                    NaN                      NaN       backbtn  trackdone   

  shuffle  offline  offline_timestamp  incognito_mode  
0    True    False         1672331958           False  
1    True    False         1672331958           False  
2    True    False         1672331960           False  
3    True    False         1672331962           False  
4    True    False         1672331963           False  

[5 rows x 21 columns]

In [22]:
# b.
interes = ['ts','ms_played',
'master_metadata_track_name', 'master_metadata_album_artist_name',
'master_metadata_album_album_name', 'episode_name',
'episode_show_name',
'reason_start', 'reason_end']
df = df[interes]
df.head()

ts  ms_played master_metadata_track_name  \
0  2022-12-29T16:39:18Z        753               Warning Sign   
1  2022-12-29T16:39:19Z       1661                    Houdini   
2  2022-12-29T16:39:21Z       1000               Fingerprints   
3  2022-12-29T16:39:22Z       1097            Waves of Nature   
4  2022-12-29T16:43:38Z     256920               Fingerprints   

  master_metadata_album_artist_name master_metadata_album_album_name  \
0                          Coldplay      A Rush of Blood to the Head   
1                 Foster The People                          Torches   
2                    Hiatus Kaiyote               Choose Your Weapon   
3                            WILLOW                     ARDIPITHECUS   
4                    Hiatus Kaiyote               Choose Your Weapon   

  episode_name episode_show_name reason_start reason_end  
0          NaN               NaN       fwdbtn     fwdbtn  
1          NaN               NaN       fwdbtn     fwdbtn  
2          NaN               NaN       fwdbtn     fwdbtn  
3          NaN               NaN       fwdbtn    backbtn  
4          NaN               NaN      backbtn  trackdone

In [23]:
# c.
df

ts  ms_played   master_metadata_track_name  \
0      2022-12-29T16:39:18Z        753                 Warning Sign   
1      2022-12-29T16:39:19Z       1661                      Houdini   
2      2022-12-29T16:39:21Z       1000                 Fingerprints   
3      2022-12-29T16:39:22Z       1097              Waves of Nature   
4      2022-12-29T16:43:38Z     256920                 Fingerprints   
...                     ...        ...                          ...   
41822  2025-07-13T15:18:31Z      29642                  Flintstones   
41823  2025-07-13T15:21:40Z     187573                   Moon River   
41824  2025-07-13T15:26:51Z      54677                   Moon River   
41825  2025-07-13T15:28:54Z     122345                   Moon River   
41826  2025-07-15T13:45:38Z          0  Moon River - Alternate Take   

      master_metadata_album_artist_name  \
0                              Coldplay   
1                     Foster The People   
2                        Hiatus Kaiyote   
3                                WILLOW   
4                        Hiatus Kaiyote   
...                                 ...   
41822                     Jacob Collier   
41823                     Jacob Collier   
41824                   Louis Armstrong   
41825                    Audrey Hepburn   
41826                     Henry Mancini   

                  master_metadata_album_album_name episode_name  \
0                      A Rush of Blood to the Head          NaN   
1                                          Torches          NaN   
2                               Choose Your Weapon          NaN   
3                                     ARDIPITHECUS          NaN   
4                               Choose Your Weapon          NaN   
...                                            ...          ...   
41822                                   In My Room          NaN   
41823                                Djesse Vol. 2          NaN   
41824                                Hello, Dolly!          NaN   
41825  Fruhstuck Bei Tiffany (Original Soundtrack)          NaN   
41826                                     Mr Lucky          NaN   

      episode_show_name reason_start reason_end  
0                   NaN       fwdbtn     fwdbtn  
1                   NaN       fwdbtn     fwdbtn  
2                   NaN       fwdbtn     fwdbtn  
3                   NaN       fwdbtn    backbtn  
4                   NaN      backbtn  trackdone  
...                 ...          ...        ...  
41822               NaN     clickrow    endplay  
41823               NaN      playbtn    endplay  
41824               NaN      playbtn    endplay  
41825               NaN      playbtn    endplay  
41826               NaN    trackdone     remote  

[41827 rows x 9 columns]

C)
El dataset tiene 41827 filas

## 2. Limpieza


a. Revise si existen filas duplicadas en su dataset. Si es así, deje solamente la última aparición de la fila duplicada en su DataFrame. ¿Cuántas filas duplicadas hay?

b. Ahora, revise si hay filas cuyo valor en `ms_played` (milisegundos reproducidos) es menor a 1. Descarte todas estas filas.

c. ¿Cuántas filas quedaron en su DataFrame después de la limpieza?

In [24]:
# a.
df.duplicated().sum()



np.int64(72)

In [25]:
df = df.drop_duplicates(keep='last')
df

ts  ms_played   master_metadata_track_name  \
0      2022-12-29T16:39:18Z        753                 Warning Sign   
1      2022-12-29T16:39:19Z       1661                      Houdini   
2      2022-12-29T16:39:21Z       1000                 Fingerprints   
3      2022-12-29T16:39:22Z       1097              Waves of Nature   
4      2022-12-29T16:43:38Z     256920                 Fingerprints   
...                     ...        ...                          ...   
41822  2025-07-13T15:18:31Z      29642                  Flintstones   
41823  2025-07-13T15:21:40Z     187573                   Moon River   
41824  2025-07-13T15:26:51Z      54677                   Moon River   
41825  2025-07-13T15:28:54Z     122345                   Moon River   
41826  2025-07-15T13:45:38Z          0  Moon River - Alternate Take   

      master_metadata_album_artist_name  \
0                              Coldplay   
1                     Foster The People   
2                        Hiatus Kaiyote   
3                                WILLOW   
4                        Hiatus Kaiyote   
...                                 ...   
41822                     Jacob Collier   
41823                     Jacob Collier   
41824                   Louis Armstrong   
41825                    Audrey Hepburn   
41826                     Henry Mancini   

                  master_metadata_album_album_name episode_name  \
0                      A Rush of Blood to the Head          NaN   
1                                          Torches          NaN   
2                               Choose Your Weapon          NaN   
3                                     ARDIPITHECUS          NaN   
4                               Choose Your Weapon          NaN   
...                                            ...          ...   
41822                                   In My Room          NaN   
41823                                Djesse Vol. 2          NaN   
41824                                Hello, Dolly!          NaN   
41825  Fruhstuck Bei Tiffany (Original Soundtrack)          NaN   
41826                                     Mr Lucky          NaN   

      episode_show_name reason_start reason_end  
0                   NaN       fwdbtn     fwdbtn  
1                   NaN       fwdbtn     fwdbtn  
2                   NaN       fwdbtn     fwdbtn  
3                   NaN       fwdbtn    backbtn  
4                   NaN      backbtn  trackdone  
...                 ...          ...        ...  
41822               NaN     clickrow    endplay  
41823               NaN      playbtn    endplay  
41824               NaN      playbtn    endplay  
41825               NaN      playbtn    endplay  
41826               NaN    trackdone     remote  

[41755 rows x 9 columns]

A)
Hay 72 filas duplicadas

In [15]:
# b.
df = df[df['ms_played'] >= 1]
df

ts  ms_played  master_metadata_track_name  \
0      2022-12-29T16:39:18Z        753                Warning Sign   
1      2022-12-29T16:39:19Z       1661                     Houdini   
2      2022-12-29T16:39:21Z       1000                Fingerprints   
3      2022-12-29T16:39:22Z       1097             Waves of Nature   
4      2022-12-29T16:43:38Z     256920                Fingerprints   
...                     ...        ...                         ...   
41821  2025-07-13T15:17:37Z     116308  Salmo de la Rosa Verdadera   
41822  2025-07-13T15:18:31Z      29642                 Flintstones   
41823  2025-07-13T15:21:40Z     187573                  Moon River   
41824  2025-07-13T15:26:51Z      54677                  Moon River   
41825  2025-07-13T15:28:54Z     122345                  Moon River   

      master_metadata_album_artist_name  \
0                              Coldplay   
1                     Foster The People   
2                        Hiatus Kaiyote   
3                                WILLOW   
4                        Hiatus Kaiyote   
...                                 ...   
41821                     Inti-Illimani   
41822                     Jacob Collier   
41823                     Jacob Collier   
41824                   Louis Armstrong   
41825                    Audrey Hepburn   

                  master_metadata_album_album_name episode_name  \
0                      A Rush of Blood to the Head          NaN   
1                                          Torches          NaN   
2                               Choose Your Weapon          NaN   
3                                     ARDIPITHECUS          NaN   
4                               Choose Your Weapon          NaN   
...                                            ...          ...   
41821                              Lugares Comunes          NaN   
41822                                   In My Room          NaN   
41823                                Djesse Vol. 2          NaN   
41824                                Hello, Dolly!          NaN   
41825  Fruhstuck Bei Tiffany (Original Soundtrack)          NaN   

      episode_show_name reason_start reason_end  
0                   NaN       fwdbtn     fwdbtn  
1                   NaN       fwdbtn     fwdbtn  
2                   NaN       fwdbtn     fwdbtn  
3                   NaN       fwdbtn    backbtn  
4                   NaN      backbtn  trackdone  
...                 ...          ...        ...  
41821               NaN    trackdone    endplay  
41822               NaN     clickrow    endplay  
41823               NaN      playbtn    endplay  
41824               NaN      playbtn    endplay  
41825               NaN      playbtn    endplay  

[41487 rows x 9 columns]

In [17]:
# c.
df

ts  ms_played  master_metadata_track_name  \
0      2022-12-29T16:39:18Z        753                Warning Sign   
1      2022-12-29T16:39:19Z       1661                     Houdini   
2      2022-12-29T16:39:21Z       1000                Fingerprints   
3      2022-12-29T16:39:22Z       1097             Waves of Nature   
4      2022-12-29T16:43:38Z     256920                Fingerprints   
...                     ...        ...                         ...   
41821  2025-07-13T15:17:37Z     116308  Salmo de la Rosa Verdadera   
41822  2025-07-13T15:18:31Z      29642                 Flintstones   
41823  2025-07-13T15:21:40Z     187573                  Moon River   
41824  2025-07-13T15:26:51Z      54677                  Moon River   
41825  2025-07-13T15:28:54Z     122345                  Moon River   

      master_metadata_album_artist_name  \
0                              Coldplay   
1                     Foster The People   
2                        Hiatus Kaiyote   
3                                WILLOW   
4                        Hiatus Kaiyote   
...                                 ...   
41821                     Inti-Illimani   
41822                     Jacob Collier   
41823                     Jacob Collier   
41824                   Louis Armstrong   
41825                    Audrey Hepburn   

                  master_metadata_album_album_name episode_name  \
0                      A Rush of Blood to the Head          NaN   
1                                          Torches          NaN   
2                               Choose Your Weapon          NaN   
3                                     ARDIPITHECUS          NaN   
4                               Choose Your Weapon          NaN   
...                                            ...          ...   
41821                              Lugares Comunes          NaN   
41822                                   In My Room          NaN   
41823                                Djesse Vol. 2          NaN   
41824                                Hello, Dolly!          NaN   
41825  Fruhstuck Bei Tiffany (Original Soundtrack)          NaN   

      episode_show_name reason_start reason_end  
0                   NaN       fwdbtn     fwdbtn  
1                   NaN       fwdbtn     fwdbtn  
2                   NaN       fwdbtn     fwdbtn  
3                   NaN       fwdbtn    backbtn  
4                   NaN      backbtn  trackdone  
...                 ...          ...        ...  
41821               NaN    trackdone    endplay  
41822               NaN     clickrow    endplay  
41823               NaN      playbtn    endplay  
41824               NaN      playbtn    endplay  
41825               NaN      playbtn    endplay  

[41487 rows x 9 columns]

c) Quedan 41487 filas en el dataframe


## 3. Transformación

Aplique las siguientes transformaciones a su DataFrame:

a. La columna `ts` (timestamp) muestra la fecha y hora de reproducción de la canción. Extraiga únicamente el año y agregue una nueva columna con este dato como número entero. ¿En qué año hubo más reproducciones?

b. El dataset contiene canciones y episodios de podcasts. Genere una nueva columna (categórica o booleana) que identifique a cada fila como "canción" o "episodio" y explique cómo realizó esta clasificación. ¿Cuántas filas hay de cada una de estas categorías? 

c. Responda:
¿Cuál es el podcast con mayor número de reproducciones en total?

In [28]:
# a.
df["ts"] = pd.to_datetime(df["ts"])
df["año"] = df["ts"].dt.year
df.head()

C:\Users\gasxd\AppData\Local\Temp\ipykernel_21476\3887812324.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["ts"] = pd.to_datetime(df["ts"])
C:\Users\gasxd\AppData\Local\Temp\ipykernel_21476\3887812324.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["año"] = df["ts"].dt.year


ts  ms_played master_metadata_track_name  \
0 2022-12-29 16:39:18+00:00        753               Warning Sign   
1 2022-12-29 16:39:19+00:00       1661                    Houdini   
2 2022-12-29 16:39:21+00:00       1000               Fingerprints   
3 2022-12-29 16:39:22+00:00       1097            Waves of Nature   
4 2022-12-29 16:43:38+00:00     256920               Fingerprints   

  master_metadata_album_artist_name master_metadata_album_album_name  \
0                          Coldplay      A Rush of Blood to the Head   
1                 Foster The People                          Torches   
2                    Hiatus Kaiyote               Choose Your Weapon   
3                            WILLOW                     ARDIPITHECUS   
4                    Hiatus Kaiyote               Choose Your Weapon   

  episode_name episode_show_name reason_start reason_end   año  
0          NaN               NaN       fwdbtn     fwdbtn  2022  
1          NaN               NaN       fwdbtn     fwdbtn  2022  
2          NaN               NaN       fwdbtn     fwdbtn  2022  
3          NaN               NaN       fwdbtn    backbtn  2022  
4          NaN               NaN      backbtn  trackdone  2022

In [29]:
df['año'].value_counts()

año
2024    18940
2023    15573
2025     7116
2022      126
Name: count, dtype: int64

El año con más reproducciones fue 2024.

In [37]:
# b.

df["es_episodio"] = df["episode_name"].isnull()
df
df = df.drop("es_podcast",)
df

TypeError: 'NoneType' object is not subscriptable

In [ ]:
# c.

El podcast con más reproducciones es XXX, con YYY reproducciones.

## 4. Variables categóricas

a. Las columnas `reason_start`y `reason_end` muestran, respectivamente, por qué se comenzó y se terminó la reproducción de una pista. Estas son columnas categóricas. Para ambas, muestre todas las categorías que aparecen en su dataset.
¿Son las mismas? Comente.

b. Vamos a crear una nueva columna llamada `skipped` (categórica o booleana), que identificará si la reproducción de la pista fue interrumpida intencionalmente por el usuario (saltada). Para esto, nos fijaremos en el valor de `reason_end` de cada fila. Considere una pista como "skipped" si su término está entre las siguientes categorías: `endplay`, `fwdbtn`, `backbtn`, y como "not skipped" en caso contrario.

c. Finalmente, responda:
¿Cuál es la canción más interrumpida (o saltada)? Responda con el nombre de la canción y el artista.

In [ ]:
# a.

In [ ]:
# b.

In [ ]:
# c.

La canción más interrumpida por el usuario es XXX.